File: perl/pe021.pl
#!/usr/bin/perl -w
use strict;

my $map;
for my $i (0 .. 9999) {
  $map->{$i} = sum_factors($i);
}

my $total;
while (my ($key, $value) = each %$map) {
  $total += $key if ((defined $map->{$value}) && ($key != $value) && ($map->{$value} == $key));
}
print $total,"\n";

sub sum_factors {
  my $number = shift;

  my $sum = 1;
  for (my $i = 2; $i <= sqrt($number); $i++) {
    if (!($number % $i)) {
      $sum += ($i + ($number / $i));
    }
  }
  return $sum;
}
  

File: c++/pe021.cpp
// This is a solution for http://projecteuler.net/problem=21

#include <cstdio>
#include <cmath>

const int SIZE = 10000;

int sum_factors(int number);

int main() {
  int sumproperdivisors[SIZE];
  int i;
  for (i = 0; i < SIZE; ++i)
    sumproperdivisors[i] = sum_factors(i);

  int total = 0;
  for (i = 0; i < SIZE; ++i)
    if ((sumproperdivisors[i] < SIZE) && (i != sumproperdivisors[i])
         && (sumproperdivisors[sumproperdivisors[i]] == i))
      total += i;

  printf("%d\n", total);
  return 0;
}

int sum_factors(int number) {
  int sum = 1;
  int i;
  for (i = 2; i <= sqrt(number); ++i)
    if (!(number % i))
      sum += (i + (number / i));

  return sum;
}
